Risu PPC64 fixes

mferst opened the pull request.

vcoracolombo committed:
ppc64: Add fpscr comparison

Signed-off-by: Víctor Colombo <victor.colombo@eldorado.org.br>
[ferst: Ignore FPSCR[FR] for now; Rebase; ]

mferst committed:
ppc64: add VSX support.

The current code can be used to test VSX instructions, but the lower
half of the first 32 VSRs are ignored. Add risugen code to
(conditionally) initialize the 128-bits of all VSX registers and C code
to compare the lower 64-bits of VSRs 0-31.

mferst committed:
ppc64: add support to prefixed insns

Works for PPC64, seems to work for m68k, but still breaks ARM.

mferst committed:
ppc64: init and compare whole ccr

mferst committed:
ppc64: initialize ri->prev_insn

mferst committed:
ppc64: initialize and compare CTR

mferst committed:
ppc64: initialize and compare VSCR and VRSAVE

mferst committed:
ppc64: use 32-bits values for each VR word

mferst committed:
ppc64: fix addg6s opcode

mferst committed:
ppc64: add !memory block to load/store insns

Also adds "$ra != $rb" and "$ra != 0" constraints to avoid the current
limitations of reg_plus_reg and reg_plus_imm implementations.

mferst committed:
ppc64: Fix LXV opcode and !memory

Unlike the other VSX instructions, the TX field of LVX is not the last
bit. Also, the immediate is multiplied by 16 to calculate EA.

mferst committed:
ppc64: Fix LXVL[L]/STXVL[L] memory block

EA is not calculated with "reg plus reg" for these instructions.
Instead, RA is the base register to access up to 16 bytes, according
to RB[0:7] value. Also, abuse the memory block to put a valid value in
RB.
--
Power ISA says that the results are "boundedly undefined" if the
contents of bits 8:63 of RB are not equal to zero. I'd expect the
hardware to ignore those bits, but we observed SIGSEGVs on hardware if
we use just reg($ra) and let any random value in RB.

mferst committed:
ppc64: fix STQ constraint

If $ra is used as one of the source registers, we'll store a value that
depends on the stack pointer in the memory block. If a load instruction
later reads this value, the register comparison will fail mysteriously.

mferst committed:
ppc64: improve write_memblock_setup

Currently, write_memblock_setup initializes 10000 bytes below the stack
pointer with zeros. However, reg_plus_reg and reg_plus_imm make the
instruction point to $r1+32 and $r1, respectively, causing access
outside the stack frame. Change write_memblock_setup to initialize the
memblock with random data, and use STDU to make $r1 point to the
beginning of the block at the end.

vcoracolombo commented on ppc64.risu:
```diff
@@ -3156,11 +3199,6 @@ XVCMPGTSP PPC64LE 111100 t:5 a:5 b:5 01001011 ax:1 bx:1 tx:1
 # format:XX3 book:I page:672 v2.06 xvcmpgtsp[.] VSX Vector Compare Greater Than SP
 XVCMPGTSPd PPC64LE 111100 t:5 a:5 b:5 11001011 ax:1 bx:1 tx:1
 
-# format:XX3 book:I page:673 v3.0 xvcmpnedp[.] VSX Vector Compare Not Equal Double-Precision
-XVCMPNEDP PPC64LE 111100 t:5 a:5 b:5 01111011 ax:1 bx:1 tx:1
-# format:XX3 book:I page:673 v3.0 xvcmpnedp[.] VSX Vector Compare Not Equal Double-Precision
-XVCMPNEDPd PPC64LE 111100 t:5 a:5 b:5 11111011 ax:1 bx:1 tx:1
-
```
So, I just noticed I sent this in this patch.

The problem with these lines is that theses instructions were removed from Power ISA. I think they were introduced in 3.0 and removed in 3.0B. Do you think we should remove them from risu?
If we remove them, I'll have to send a separated patch, because it should not be in this commit

mferst replied:
This is also the case for LMDX, it was present in the first version of PowerISA v3.0 but removed on PowerISA v3.0b. I think we can keep them commented out, just like we do with other instructions that cannot be tested like SC[V], TRAP, etc.

mferst committed:
ppc64: comment out insns that can't be tested with risu

ADDPCIS uses the program counter, which depends on where the image was
loaded. DARN is non-deterministic by definition, we can't compare the
results between executions.

vcoracolombo committed:
ppc64: comment out insns removed in PowerISA v3.0b

[ferst: LDMX; comment instead of removal]

vcoracolombo committed:
ppc64: Add fr*p constraints in ppc64.risu

Some instructions use a pair of fp registers, represented by fr*p
instead of just fr* (e.g. frap, frbp). These have the special property
that they have to be an even-odd pair, starting at the even one.
Otherwise, the instruction is malformed

This patch adds constraints to these instructions to garantee the
selected registers are even.

Signed-off-by: Víctor Colombo <victor.colombo@eldorado.org.br>

mferst committed:
ppc64: some POWER10 insns patterns

Commit:
ppc64: more POWER10 insns

mferst committed:
ppc64: prefixed insns examples

Commit:
ppc64: MMA prefixed insns patterns

Commit:
ppc64: add plxsd/pstxsd and plxssp/pstxssp
